vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java

```diff
@@ -400,6 +408,395 @@ public ElementQuery<T> withoutTheme(String theme) {
+        return withPlaceholder(placeholder, String::equals);
+    }
+
+    /**
+     * Requires the element's placeholder to partially match the given text value.
+     * <p>
+     * For exactly matching the placeholder, see
+     * {@link #withPlaceholder(String)}.
+     *
+     * @param text
+     *          the text to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholder(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholderContaining(String text) {
+        return withPlaceholder(text, String::contains);
+    }
+
+    /**
+     * Requires the element's caption (i.e., label, placeholder, or text label)
+     * to satisfy the given comparison with the supplied text.
+     * <p>
+     * If the element has a label, the label value is used.
+     * If the element can have a label, but its value is empty and it has a placeholder,
+     * the placeholder value is used.
+     * If the element supports neither labels nor placeholders
+     * but does support a caption via its text (such as a button),
+     * the text is used.
+     * <p>
+     * For matching a caption exactly, see {@link #withCaption(String)},
+     * and for matching a caption partially, see {@link #withCaptionContaining(String)}.
+     * <p>
+     * This method can be used for other caption matching needs,
+     * such as performing a case-insensitive match:
+     * {@code withCaption("name", String::equalsIgnoreCase)}
+     * <br> a caption-ending match:
+     * {@code withCaption(" Name", String::endsWith)}
+     * <br> or a regular expression match:
+     * {@code withCaption("(First|Last) Name", String::matches)}
+     *
+     * @param text
+     *          the text to compare with the caption; not null
+     * @param comparison
+     *          the comparison to use when comparing the caption
+     *          against the supplied text
+     *          (i.e., {@code comparison.test(elementCaption, text)}; not null
+     * @return this element query instance for chaining
+     *
+     * @see #withCaption(String)
+     * @see #withCaptionContaining(String)
+     */
+    @SuppressWarnings("java:S3776")  // cognitive complexity > 15
+    public ElementQuery<T> withCaption(String text, BiPredicate<String, String> comparison) {
+        Objects.requireNonNull(text, NULL_TEXT_MSG);
+        Objects.requireNonNull(comparison, NULL_COMPARISON_MSG);
+        return withCondition(element -> {
+            // special case when text is empty,
+            // so if they both exist, both label and placeholder must be empty
+            if (text.isEmpty() &&
+                    element instanceof HasLabel hasLabel &&
+                    element instanceof HasPlaceholder hasPlaceholder) {
+                var label = hasLabel.getLabel();
+                var placeholder = hasPlaceholder.getPlaceholder();
+                return label.isEmpty() && placeholder.isEmpty();
+            }
```

Comment:
This special case must be documented, since it might not be obvious to a developer.